File: src/main/python/Day17.py
import Utils as Utils

def isInside(x, y, area):
    x_min, y_min, x_max, y_max = area
    return x >= x_min and x < x_max and y >= y_min and y < y_max

def hitTarget(vx, vy, area):
    x_min, y_min, x_max, y_max = area
    x = 0
    y = 0
    ly = 0
    while True:
        if isInside(x, y, area):
            return ly
        
        if y < y_min or x > x_max:
            return -1
        
        x += vx
        y += vy
        ly = max(ly, y)
        vx -= 1
        vy -= 1
        if vx < 0:
            vx = 0

@Utils.time_it_and_evaluate
def partOne(area):
    largest_y = 0
    for y in range(-100, 100):
        for x in range(area[2]):
            largest_y = max(largest_y, hitTarget(x, y, area))
    return largest_y

@Utils.time_it_and_evaluate
def partTwo(area):
    times = 0
    for y in range(-100, 100):
        for x in range(area[2]):
            times += 1 if hitTarget(x, y, area) >= 0 else 0
    return times

if __name__ == '__main__':
    line = Utils.readAll(2021, 'day17').strip()
    parts = line[13:].split(', ')
    xparts = parts[0][2:].split("..")
    yparts = parts[1][2:].split("..")
    area = (int(xparts[0]), int(yparts[0]), int(xparts[1]) + 1, int(yparts[1]) + 1)
    
    print('Day 17')
    partOne(area)
    partTwo(area)

File: src/main/java/me/hardcoded/aoc2021/Day17.java
package me.hardcoded.aoc2021;


import java.awt.Rectangle;

import me.hardcoded.util.Utils;

public class Day17 {
	public static void main(String[] args) throws Exception {
		String line = Utils.readAll(2021, "day17").strip();

		String[] parts = line.substring(13).split(", ");
		String[] xparts = parts[0].substring(2).split("\\.\\.");
		String[] yparts = parts[1].substring(2).split("\\.\\.");
		int xMin = Integer.parseInt(xparts[0]);
		int xMax = Integer.parseInt(xparts[1]) + 1;
		int yMin = Integer.parseInt(yparts[0]);
		int yMax = Integer.parseInt(yparts[1]) + 1;
		Rectangle area = new Rectangle(xMin, yMin, xMax - xMin, yMax - yMin);
		
		Utils.printf("Day 17\n");
		Utils.printf("PartOne: %d\n", partOne(area, xMax));
		Utils.printf("PartTwo: %d\n", partTwo(area, xMax));
	}
	
	public static int hitTarget(int vx, int vy, Rectangle rect) {
		int x = 0;
		int y = 0;
		int ly = 0;
		while(true) {
			if(rect.contains(x, y)) {
				return ly;
			}
			
			if(y < rect.y || x > rect.x + rect.width) {
				return -1;
			}
			
			x += vx--;
			y += vy--;
			ly = Math.max(ly, y);
			if(vx < 0) vx = 0;
		}
	}
	
	// Solve: 14 min
	public static long partOne(Rectangle area, int xMax) throws Exception {
		int largest_y = 0;
		for(int y = -1000; y < 1000; y++) {
			for(int x = 0; x < xMax; x++) {
				largest_y = Math.max(largest_y, hitTarget(x, y, area));
			}
		}
		
		return largest_y;
	}
	
	// Solve: 6 min
	public static long partTwo(Rectangle area, int xMax) throws Exception {
		int times = 0;
		for(int y = -1000; y < 1000; y++) {
			for(int x = 0; x < xMax; x++) {
				times += (hitTarget(x, y, area) >= 0) ? 1:0;
			}
		}
		
		return times;
	}
}
